mgmt: mcumgr: Rework event callback system

nordicjm opened the pull request:
Reworks the event callback system to use a linked list to allow for chained handlers and support passing a status back to the handler to indicate if the request should be rejected or allowed.

- [X] Add slist callback system
- [X] Properly document callback struct
- [X] Convert existing mgmt_evt() calls to new system
- [X] Convert other event notifications to use new system in fs_mgmt
- [X] Convert other event notifications to use new system in img_mgmt
- [X] Convert other event notifications to use new system in os_mgmt
- [X] ~~Convert other event notifications to use new system in shell_mgmt (if any)~~
- [X] ~~Convert other event notifications to use new system in stat_mgmt (if any)~~
- [X] ~~Convert other event notifications to use new system in zephyr_basic (if any)~~
- [X] Add Kconfigs to enable/disable feature
- [X] Size comparison between old/new system
- [X] Add documentation
- [X] Add release notes
- [X] Add migration guide
- [X] Remove old callback code

Fixes #47274
Fixes #45891
Fixes #50923

Size comparison (smp_svr, registers handler, has callback functions that just outputs the provided variables to the LOG):
Old:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      159140 B       256 KB     60.71%
            SRAM:       56980 B       448 KB     12.42%
```
New (all hooks enabled):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      158804 B       256 KB     60.58%
            SRAM:       56964 B       448 KB     12.42%
```
New (all hooks and notification system disabled):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      158176 B       256 KB     60.34%
            SRAM:       56956 B       448 KB     12.42%
```

nordicjm committed:
mgmt: mcumgr: Rework event callback system

Reworks the event callback system to use a linked list to allow for
chained handlers and support passing a status back to the handler to
indicate if the request should be rejected or allowed. This also
removes the old base callback functionality.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
mgmt: mcumgr: os_mgmt: Switch to new event callback system

Switches to the new event callback system for the os_mgmt
functionality and removes the old code.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
mgmt: mcumgr: Add access denied error code

Adds an access denied error code that can be used to signal to the
mcumgr client that the requested access to a specific resource or
command/functionality has been denied.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

de-nordic commented on subsys/mgmt/mcumgr/lib/mgmt/src/mgmt.c:
```diff
@@ -65,30 +70,54 @@ mgmt_register_group(struct mgmt_group *group)
+	(void)sys_slist_find_and_remove(&mgmt_callback_list, &callback->node);
+}
+
+bool mgmt_notify_callbacks(uint32_t event, uint16_t group, uint8_t id, void *data,
```
I think that at least 3 first parameters should be part of single "context" object. These things will be the same through processing of single group, loading them separately each time you want to send notification will be overkill.
Instead of returning 'bool' status we should define some specific error set that will be returned and then translated in lower level.

nordicjm replied:
I was keeping them like this to aid migration of the existing fs_mgmt API. In terms of returning a status code, my idea for that was to use the provided struct, for messages that support it, because the vast majority of callbacks have no status on influence on running other callbacks, at present it's just the fs_mgmt file read/write callback and img_mgmt chunk callback that do checks on the return code

de-nordic replied:
> I was keeping them like this to aid migration of the existing fs_mgmt API. 

You have already announced rework, so you are free to rework it completely.
We are loosing compatibility anyway, so why bother?
Instead of using the "coordinates" to identify callback point you may just go for two:
1) Group/mcumgr subsystem
2) Notification point

These can be encoded into single 32 bit event code easily.

For example (naming and numbers may differ)
```
// 8 codes reserved main areas mgmt (group registration, etc), SMP protocol, buffer allocations
#define MCUMGR_MGMT_NOTIFY_ID(area, point) (area << 8 | point)
#define MCUMGR_MGMT_NOTIFY_AREA_MGMT 1
#define MCUMGR_MGMT_NOTIFY_MGMT_REGISTER_GRP MCUMGR_MGMT_NOTIFY_ID(MCUMGR_MGMT_NOTIFY_AREA_MGMT, 10)
....
// 16 codes, starting at 16 reserved for transports. Will there be more?
#define MCUMGR_MGMT_NOTIFY_AREA_TRANSPORT 9
....
```

Then it is enough to do:
```
if (point == MCUMGR_MGMT_NOTIFY_ID(MCUMGR_MGMT_...AREA..., MCUMGR_MGMT_....))
```
or
```
switch (id)
{
case MCUMGR_MGMT_NOTIFY_ID(MCUMGR_MGMT_...AREA..., MCUMGR_MGMT_....):
...
```

To process events. 

You have chance to design single callback infrastructure to serve  informative/review/security purposes? Why not take time to do so?

> In terms of returning a status code, my idea for that was to use the provided struct, for messages that support it, because the vast majority of callbacks have no status on influence on running other callbacks (...)

Things returned by callbacks have to be clearly defined so that user would know what to return and how does this affect mcumgr code, this would require defining "callback points" to identify and pair them with values. MCUmgr code can only react to what it knows and we do not want everyone to expand logic in MCUmgr, to serve notifications.

For example (naming is just for example):
event point `MCUMGR_MGMT_NOTIFY_GRP_FS_DOWNLOAD_START` expects one of:
```
 MCUMGR_MGMT_NOTIFY_FS_BUSY -- can not handle processing now, file will not be downloaded, proper BUSY response will be reported to client;
 MCUMGR_MGMT_NOTIFY_FS_OK -- will process download;
 MCUMGR_MGMT_NOTIFY_FS_UNAVAILABLE -- pretend file is not there;
 MCUMGR_MGMT_NOTIFY_FS_ACCESS_DENIED -- not authorized or missing authorization token;
 MCUMGR_MGMT_NOTIFY_FS_PM_DEVICE_DOWN -- power management failed to bring device up, used when aggressive power management is used and notification functions are used to power up/down devices at various points of download.
All other responses, for security, are treated as MCUMGR_MGMT_NOTIFY_FS_ACCESS_DENIED.
```

This really needs thinking over what we expect of it and reworking.

nordicjm replied:
> These can be encoded into single 32 bit event code easily.

That system is very limiting and not usable, it assumes a maximum of 8 mgmt groups, there are users of mcumgr that already have in excess of 20 groups in a single product, so 8 is not sufficient.

> Things returned by callbacks have to be clearly defined so that user would know what to return and how does this affect mcumgr code, this would require defining "callback points" to identify and pair them with values. MCUmgr code can only react to what it knows and we do not want everyone to expand logic in MCUmgr, to serve notifications.

Not a fan of this at all, this then introduces new return codes specific for each event. If a status is going to be returned, it would need to be one of `MGMT_ERR_*` (which needs new error codes adding)

de-nordic replied:
> > These can be encoded into single 32 bit event code easily.
> 
> That system is very limiting and not usable, it assumes a maximum of 8 mgmt groups, there are users of mcumgr that already have in excess of 20 groups in a single product, so 8 is not sufficient.

It is an example, you can take 24 bits to encode groups/transport/mgmt subsystem and 8 bit to encode event points, even though this should be overkill, because it allows to put points in 255 lines.

> 
> Not a fan of this at all, this then introduces new return codes specific for each event. If a status is going to be returned, it would need to be one of `MGMT_ERR_*` (which needs new error codes adding)

Then design common code set like `ERRNO` does, and let users extend them for the purpose of their groups.

nordicjm replied:
8 bits means a maximum of 8 callbacks for a whole mgmt group, that is not sufficient. The mgmt group ID itself is 16 bits, so that field cannot be any less than 16 bits for that alone (will rework it to use the group ID directly rather than a bitmask for that part), can use a 32 bit field for it and designate the first few bits to determine the type. Stuffing all fields together makes no sense anyway, it's just going to use more flash to stuff the bits and then more flash for every callback function where it unstuffs the bits which can instead be provided as 2 direct variables instead.

de-nordic replied:
> 8 bits means a maximum of 8 callbacks for a whole mgmt group, that is not sufficient. 

8 bits, means 255 identifiers, not counting 0, for callback points. Then you have 24 bits where you can use 8 bit to identify what additional 16 bits mean: groups, transports, etc.

Why would you use bit flags flags? It is not convenient to mark where did notification come from with flag. You can exactly identify path by giving it direct identifier.

> Stuffing all fields together makes no sense anyway, it's just going to use more flash to stuff the bits and then more flash for every callback function where it unstuffs the bits which can instead be provided as 2 direct variables instead.

I am not sure about this, note that calling function requires putting all the parameters in registers anyway, which is ops that go to flash, and storing some of them on stack which is ops in flash and space on stack.

de-nordic replied:
> It's very easy from a code perspective to enable/disable/check which bits are enabled. Not using bitmask values means having to then have individual lists for every event to register for, if a function wants to register for 1 fs_mgmt event and 3 img_mgmt events, it now needs to register 4 times separately, and the callback system needs to have n (n=events, where in this case n=4) sets of handler details to check through instead of (n=handlers, where in this case n=1, or n=2 if the bitmasks for each group are different). In a system with 4 callbacks each checking 1 event from a single group, with a bitmask value, n=4, with separate events, n=4 In a system with 4 callbacks each checking 3 events from a single group, with a bitmask value, n=4, with separate events, n=12 With 32 or 64-bit size, in the current system (assuming it is re-written so that the bitmask values are reused for each mgmt group) that allows for 32 or 64 callbacks per group with a linear access time and flat variable size, switching to incrementing values instead would allow for 255 events per group with an 8-bit variable size **per registered event** and increase the access time (though I think we can disregard the access time for these)

No. You want to dump as much work on user code as possible.
You only need one list: of mcumgr callbacks from various subsystems of an app.
There should be Kconfig options to disable notify function from being invoked, in case your app is not interested at some point at all.
The app callback should be responsible for reacting to codes, even maybe having some state machine to take care of it.

For example app callback can have something like this:
```
callback(id, prev_callback_result, arg) {
    ...
    if (just_let_it_go) {
        return OK;
    }

    if (i_dont_like_fs_now && ((MCUMGR_MGMT_NOTIFY_AREA_GET(id) == MCUMGR_MGMT_NOTIFY_AREA_FS)) {
        return NO;
    }

    if (MCUMGR_MGMT_NOTIFY_AREA_GET(id) == MCUMGR_MGMT_NOTIFY_AREA_FS) {
        switch(MCUMGR_MGMT_NOTIFY_POINT_GET(id)) {
        case MCUMGR_MGMT_NOTIFY_FS_DOWNLOAD:
        {
            fs_notify_ctxt *fnc = arg;
            if (arg->token == current_token && is_download_path_allowed(current_token, fnc->path)) {
                if (offset == 0) {
                    global_state_busy = true;
                } else {
                    global_state_busy = false;
                }
                return OK;
            } else {
                return ACCESS_DENIED;
            }
        }
        default:
            ... //some def action for other points;
        } // end switch;
    if (MCUMGR_MGMT_NOTIFY_AREA_GET(id) == MCUMGR_MGMT_NOTIFY_AREA_OS && 
        MCUMGR_MGMT_NOTIFY_POINT_GET(id)  == MCUMGR_MGMT_NOTIFY_POINT_RESET_ASK) {
        if (global_state_busy == true) {
            return NO;
        }
    }
    return prev_callback_result;
 }
```

I am not even sure we need slist for that, because we may actually make callback registration function to return previous callback and make the user code to call the returned pointer. Little bit risky, but probably more efficient and flexible.

nordicjm replied:
This is at odds with the style of callbacks in zephyr https://docs.zephyrproject.org/latest/connectivity/networking/api/net_mgmt.html#listening-to-network-events
It does not make sense to invoke a callback (or multiple) every time a chunk of data is received in img_mgmt for a bunch of handlers that do not need it, that's 1024 pointless callbacks per handler for a 256KB image upload with 256 byte MTU.

nordicjm replied:
> This will also be called every time callback gets entered, so 1024 times:

The callback system will be called 1024 times, the correct handler will be called 1024 times. If there's 10 handlers with only 1 being needed for the check, the overhead of that without using a bitmask is 9 * 1024 = 9216. That overhead is huge.

> is up to callback

That is down to the implementation, for notifications for file access, if there are 10 handlers and one denies access to the file, there is no point in running the other handlers because one of those has already denied access to it, the callback is specifically for checking if file access is allowed or not, but there may be a callback that is used to inform of something, so even if a handler fails it will still call the others.

de-nordic replied:
> > This will also be called every time callback gets entered, so 1024 times:
> 
> The callback system will be called 1024 times, the correct handler will be called 1024 times. If there's 10 handlers with only 1 being needed for the check, the overhead of that without using a bitmask is 9 * 1024 = 9216. That overhead is huge.

You have missed this one (https://github.com/zephyrproject-rtos/zephyr/pull/50578#discussion_r987902648): 

Then we can have per-group/per-subsystem callback, but we should not block app from registering the same callback function in all of them and take the consequences, which means processing them in single state machine.

> That is down to the implementation, for notifications for file access, if there are 10 handlers and one denies access to the file, there is no point in running the other handlers because one of those has already denied access to it, the callback is specifically for checking if file access is allowed or not, but there may be a callback that is used to inform of something, so even if a handler fails it will still call the others. As a real world example you might have 2 handlers, one for access checking, one for rewriting the path, why would the path be rewritten if the access check fails? It is literally a complete waste of time, processing and power.

That is not your decision, that is not my decision, that is not decision MCUmgr should take. MCUmgr should get the code and behave as the code for specific point defines. Callbacks should only report something or return specific code according to logic within them. It is on the callback to change error level, if desired; callback may also decide whether it will call the next one, or allow the next one to be called, or just end the sequence and return code.

> As a real world example you might have 2 handlers, one for access checking, one for rewriting the path, why would the path be rewritten if the access check fails? It is literally a complete waste of time, processing and power.

If callback does it, then it does it and it is beyond us. In your real world example callback removing file may be called first, before the one that gives access; it may also, due to programmers error actually override the one that would give access with denial, because it failed to find the file. 
Neither of proposed solutions warranties order of callback execution. The one where callback decides, by calling next in order, does provide limited ability to callback to decide (at cost of stack), because callback may at any point call next in queue and decide what to do with returned value.

nordicjm replied:
Callback system updated to use zephyr callback design

carlescufi reviewed (changes requested):
Missing release notes. Please add a migration guide as well, at least a couple of sentences.

nordicjm commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/include/fs_mgmt/fs_mgmt.h:
```diff
@@ -19,38 +20,28 @@ extern "C" {
- * @return 0 to allow read/write, MGMT_ERR_[...] code to disallow read/write.
+/* Structure provided in the MGMT_EVT_OP_FS_MGMT_FILE_ACCESS notification callback: This callback
+ * function is used to notify the application about a pending file read/write request and to
+ * authorise or deny it. If the handler returns true then the next registered notification handler
+ * will be checked, if a handler returns false then the check will be aborted and access to the
+ * file will be denied.
```
Update comment

nordicjm commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/include/img_mgmt/img_mgmt.h:
```diff
@@ -104,6 +105,17 @@ struct img_mgmt_upload_action {
 
+/* Structure provided in the MGMT_EVT_OP_IMG_MGMT_DFU_CHUNK notification callback: This callback
+ * function is used to notify the application about a pending firmware upload packet from a client
+ * and authorise or deny it. If the handler returns true then the next registered notification
+ * handler will be checked, if a handler returns false then the check will be aborted and access
+ * to the file will be denied.
```
Update comment

nordicjm commented on subsys/mgmt/mcumgr/lib/mgmt/include/mgmt/mgmt.h:
```diff
@@ -90,24 +103,39 @@ struct mgmt_evt_op_cmd_status_arg {
 	int status;
```
Add group/ID

nordicjm commented on subsys/mgmt/mcumgr/lib/mgmt/src/mgmt.c:
```diff
@@ -65,30 +70,63 @@ mgmt_register_group(struct mgmt_group *group)
+	int return_rc = MGMT_ERR_EOK;
+
+	/*
+	 * Search through the linked list for entries that have registered for this event and
+	 * notify them, if a function returns false and stop_on_fail is set to true, will abort
+	 * calling additional handlers as the callback function has indicated it is rejecting the
+	 * request (this is implementation-defined behaviour).
```
Update comment

nordicjm committed:
doc: device_mgmt: callbacks: Add initial documentation

Adds the initial documentation describing the mcumgr callback
system and how to use it, including API reference and samples.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 3.3: Add mcumgr stable API change

Adds details of the stable API change to mcumgr's callback system,
including a link to the migration guide.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
> Missing release notes. Please add a migration guide as well, at least a couple of sentences.

Documentation added

de-nordic reviewed (dismissed):
Generally looks neat. 
There are problems with:
 * blocking extensibility of event coverage within in-tree code for mcumgr;
 * forcing user to register for each event independently;
 * collision user/other unit event codes

nordicjm commented on include/zephyr/mgmt/mcumgr/callbacks.h:
```diff
@@ -0,0 +1,177 @@
+	MGMT_EVT_OP_OS_MGMT_RESET,
+
+	/** Start ID to be used for custom user events. */
+	MGMT_EVT_OP_USER_CUSTOM_START,
```
Right I get what you mean about combining the ID, I will rework it. I don't understand the part about being used outside though, the events would be compiled and used for an application so the order would not matter if it changed (which it should not do anyway, new elements would be appended), or is that for the use case of having an event handler which outputs the callback details out to e.g. a UART for an external application to parse?

nordicjm replied:
Can we gut and repurpose these IDs since they are unimplemented in the zephyr version?
```
#define MGMT_GROUP_ID_CONFIG    3
#define MGMT_GROUP_ID_CRASH     5
#define MGMT_GROUP_ID_SPLIT     6
#define MGMT_GROUP_ID_RUN       7
```
Would need 2 IDs, one for base SMP and other for all groups, alternatively could add them as 62 and 63 but seems like these unused IDs could be reused instead.

de-nordic replied:
> Can we gut and repurpose these IDs since they are unimplemented in the zephyr version?
> 
> ```
> #define MGMT_GROUP_ID_CONFIG    3
> #define MGMT_GROUP_ID_CRASH     5
> #define MGMT_GROUP_ID_SPLIT     6
> #define MGMT_GROUP_ID_RUN       7
> ```
> 
> Would need 2 IDs, one for base SMP and other for all groups, alternatively could add them as 62 and 63 but seems like these unused IDs could be reused instead.

I am reluctant to remove them. I am quite sure that we will not be using the MGMT_GROUP_ID_SPLIT , but not so sure about others.
Maybe we just need to shift all groups by 4 (3), to reserve 15 (8) first codes for internal stuff like SMP, MGT, transports(?), and leave the rest of space for groups.
I do not think that we will use more than 255 points per event source, so we have more than 16 bits to allocate for these sources (groups, transports, internals).

nordicjm commented:
Updated to add group and allow registering for all events, or all events of a specific group. Documentation has not been updated, @de-nordic can you check this looks good?

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/Kconfig:
```diff
@@ -84,6 +84,22 @@ config IMG_MGMT_FRUGAL_LIST
 	  a device but requires support in client software, which has to default omitted values.
 	  Works correctly with mcumgr-cli.
 
+config IMG_MGMT_UPLOAD_CHECK_HOOK
```
Please rename this as `MGMT_MCUMGR_GRP_IMG_UPLOAD_CHECK_HOOK`.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/Kconfig:
```diff
@@ -84,6 +84,22 @@ config IMG_MGMT_FRUGAL_LIST
+	  registered callbacks to check with the user application if an upload should be accepted
+	  or rejected.
+
+config IMG_MGMT_STATUS_HOOKS
```
Please rename this as `MGMT_MCUMGR_GRP_IMG_STATUS_HOOKS`

de-nordic commented on include/zephyr/mgmt/mcumgr/callbacks.h:
```diff
@@ -0,0 +1,219 @@
+ * values.
+ */
+enum mgmt_callback_groups {
+	MGMT_CB_GROUP_ALL = 0,
```
`MGMT_EVT_CB` or `MGMT_EVNT_GRP`, you have basically created the `MGMT_EVT` namespace, so lets use it.

de-nordic commented:
> Updated to add group and allow registering for all events, or all events of a specific group. Documentation has not been updated, @de-nordic can you check this looks good?

@nordicjm  This looks great!
I do not know whether you want to address dynamic group allocation, as I have proposed in comment, or we can postpone it to next PR. I would go for postponing to get the base in, as it already looks great.

nordicjm commented:
@de-nordic will postpone it for a future v2, will add a card

Damian-Nordic commented on include/zephyr/mgmt/mcumgr/img_mgmt_callbacks.h:
```diff
@@ -0,0 +1,46 @@
+
+	/** Upload request information */
+	struct img_mgmt_upload_req *req;
+} __packed;
```
Why does this need to be packed?

nordicjm replied:
On Cortex M0 which does not support unaligned memory access, if this is not packed then it faults due to it accessing non-word boundary elements

Damian-Nordic replied:
Hmm. And why the unaligned memory access happens? `img_mgmt_upload_check` seems to be only instantiated on stack and then passed to the handlers. I'm not sure why it causes a problem.

nordicjm replied:
The original issue is here: https://github.com/zephyrproject-rtos/zephyr/issues/49066

Essentially the code accesses created objects which references objects inside of the req object. I don't have an M0 to test on if this is needed for this specific struct or not

Damian-Nordic replied:
The original issue seemed to be casting unaligned memory to the non-packed type (so the code accessing that type might incorrectly assume that the data members are aligned). Here I can't see such a cast, but I don't have time to verify it so nevermind :)

Damian-Nordic commented on subsys/mgmt/mcumgr/lib/mgmt/src/mgmt.c:
```diff
@@ -64,30 +73,70 @@ mgmt_register_group(struct mgmt_group *group)
+	 * Search through the linked list for entries that have registered for this event and
+	 * notify them, the first handler to return an error code will have this error returned
+	 * to the calling function, errors returned by additional handlers will be ignored. If
+	 * all handlers return  Access will be allowed so long as all notification handlers
```
```suggestion
	 * all handlers return  Access will be allowed so long as all notification handlers
```
Looks like part of the sentence hasn't been removed after editing.

Damian-Nordic reviewed (commented):
I have some questions and doc suggestions, but I like the change very much :)

nordicjm commented on doc/services/device_mgmt/mcumgr_callbacks.rst:
```diff
@@ -0,0 +1,282 @@
+    int my_function(uint32_t event, bool failed, bool *abort_more, void *data,
+                    size_t data_size)
+    {
+        if (event == MGMT_EVT_OP_CMD_DONE) {
```
Yes, it just demonstrates how to structure the base handler

nordicjm commented on include/zephyr/mgmt/mcumgr/callbacks.h:
```diff
@@ -0,0 +1,221 @@
+ * This callback function is used to notify an application or system about a mcumgr mgmt event.
+ *
+ * @param event		MGMT_EVT_OP_[...].
+ * @param failed	Will be set to true if a previous handler has already returned a failure
```
It is registration order dependent, yes. Updated.

Damian-Nordic replied:
And regarding my question: "What is an example scenario where passing status code between handlers is useful?", do we anticipate any specific scenario or we're adding more flexibility just in case?

nordicjm replied:
For me, no, I don't see the usage for it. For example in the file access callback, if there are 5 registered handlers and the 3rd one denies it, I don't see the point in handler 4 and 5 being called. As for if others see use in it I do not know, Dominik's idea is that handlers that deal with getting notifications only might want it e.g. a logger

nordicjm commented:
Rebased

de-nordic commented on include/zephyr/mgmt/mcumgr/callbacks.h:
```diff
@@ -0,0 +1,221 @@
+#include <mgmt/mgmt.h>
+
+#ifdef CONFIG_MCUMGR_CMD_FS_MGMT
+#include <zephyr/mgmt/mcumgr/fs_mgmt_callbacks.h>
```
zephyr/mgmt/mcumgr/grp/fs_mgmt/fs_mgmt_callback.h

de-nordic commented on include/zephyr/mgmt/mcumgr/callbacks.h:
```diff
@@ -0,0 +1,221 @@
+#endif
+
+#ifdef CONFIG_MCUMGR_CMD_IMG_MGMT
+#include <zephyr/mgmt/mcumgr/img_mgmt_callbacks.h>
```
zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_callback.h

de-nordic commented on include/zephyr/mgmt/mcumgr/callbacks.h:
```diff
@@ -0,0 +1,221 @@
+ * MGMT event callback group IDs. Note that this is not a 1:1 mapping with MGMT_GROUP_ID_[...]
+ * values.
+ */
+enum mgmt_callback_groups {
```
Maybe stick to `cb` for type definitions.

de-nordic commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -101,6 +101,30 @@ config MGMT_MAX_DECODING_LEVELS
 	  size of cbor_nb_reader structure by zcbor_state_t size per
 	  one unit selected here.
 
+config MGMT_MCUMGR_MGMT_NOTIFICATION_HOOKS
```
Sorry for asking to change this again, but I have agreed that the prefix `MGMT_` is too much, and `MCUMGR_` is distinctive enough.

de-nordic commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -101,6 +101,30 @@ config MGMT_MAX_DECODING_LEVELS
 	  one unit selected here.
 
+config MGMT_MCUMGR_MGMT_NOTIFICATION_HOOKS
+	bool "MCUMGR notification hook support"
```
MCUMGR->MCUmgr

de-nordic commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -101,6 +101,30 @@ config MGMT_MAX_DECODING_LEVELS
+config MGMT_MCUMGR_MGMT_NOTIFICATION_HOOKS
+	bool "MCUMGR notification hook support"
+	help
+	  With this enabled, applications and parts of code can register for MCUMGR event
```
MCUMGR-MCUmgr

de-nordic commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -101,6 +101,30 @@ config MGMT_MAX_DECODING_LEVELS
+	  function that the request should be denied, for informal-only notifications or
+	  acceptable, true must be returned by all the registered notification handlers.
+
+config MGMT_MCUMGR_SMP_COMMAND_STATUS_HOOKS
```
Sorry for asking to change this again, but I have agreed that the prefix `MGMT_` is too much, and `MCUMGR_` is distinctive enough.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/Kconfig:
```diff
@@ -125,8 +125,9 @@ config FS_MGMT_PATH_SIZE
 	  and download commands.
 
-config FS_MGMT_FILE_ACCESS_HOOK
+config MGMT_MCUMGR_GRP_FS_FILE_ACCESS_HOOK
```
Sorry for asking to change this again, but I have agreed that the prefix `MGMT_` is too much, and `MCUMGR_` is distinctive enough.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/Kconfig:
```diff
@@ -84,6 +84,22 @@ config IMG_MGMT_FRUGAL_LIST
 	  a device but requires support in client software, which has to default omitted values.
 	  Works correctly with mcumgr-cli.
 
+config MGMT_MCUMGR_GRP_IMG_UPLOAD_CHECK_HOOK
```
Sorry for asking to change this again, but I have agreed that the prefix `MGMT_` is too much, and `MCUMGR_` is distinctive enough.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/Kconfig:
```diff
@@ -84,6 +84,22 @@ config IMG_MGMT_FRUGAL_LIST
+	  registered callbacks to check with the user application if an upload should be accepted
+	  or rejected.
+
+config MGMT_MCUMGR_GRP_IMG_STATUS_HOOKS
```
Sorry for asking to change this again, but I have agreed that the prefix `MGMT_` is too much, and `MCUMGR_` is distinctive enough.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/Kconfig:
```diff
@@ -19,9 +19,10 @@ config OS_MGMT_RESET_MS
 	  response to be delivered.
 
-config OS_MGMT_RESET_HOOK
+config MGMT_MCUMGR_GRP_OS_OS_RESET_HOOK
```
Sorry for asking to change this again, but I have agreed that the prefix `MGMT_` is too much, and `MCUMGR_` is distinctive enough.

de-nordic commented on include/zephyr/mgmt/mcumgr/callbacks.h:
```diff
@@ -0,0 +1,221 @@
+ * @return		MGMT_ERR_[...] either MGMT_ERR_EOK if all handlers returned it, or the
+ *			error code of the first handler that returned an error.
+ */
+int32_t mgmt_notify_callbacks(uint32_t event, void *data, size_t data_size);
```
Would you consider changing order to `mgmt_callback_notify`, `mgmt_callback_register` and `mgmt_callback_unregister`.
This would crate `mgmt_callback` namespace and make the functions look like they work around the same area.

nordicjm committed:
mgmt: mcumgr: fs_mgmt: Switch to new event callback system

Switches to the new event callback system for the fs_mgmt
functionality and removes the old code.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
mgmt: mcumgr: img_mgmt: Switch to new event callback system

Switches to the new event callback system for the img_mgmt
functionality and removes the old code.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
Issues addressed and rebased

de-nordic reviewed (dismissed):
Looks OK. Thanks!

carlescufi commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_callbacks.h:
```diff
@@ -0,0 +1,46 @@
+
+	/** Upload request information */
+	struct img_mgmt_upload_req *req;
+} __packed;
```
Please remove this, it is not necessary.

de-nordic commented:
@Damian-Nordic @carlescufi Please take a look.

carlescufi commented:
@nordicjm please rebase

nordicjm committed:
mgmt: mcumgr: fs_mgmt: Update hash/checksum supported command

Updates the hash/checksum supported command to use the new command
structure.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
